libnmstate/state.py

```diff
@@ -444,21 +466,38 @@ def merge_routes(self, other_state):
           - Self iface explicitly specified (regardless of routes)
         - Self absent routes overwrite other routes state.
           - Support wildcard for matching the absent routes.
+        - Other routes next hop to DHCP/Autoconf interface will be discarded.
         """
         other_routes = set()
         for ifname, routes in other_state.config_iface_routes.items():
             if (
                 ifname in self.config_iface_routes
                 or self._is_interface_routable(ifname, routes)
             ):
-                other_routes |= set(routes)
+                for route in routes:
+                    if not self._route_is_next_hop_to_auto_iface(
+                        route.to_dict()
+                    ):
+                        other_routes.add(route)
 
         self_routes = {
             route
             for routes in self.config_iface_routes.values()
             for route in routes
             if not route.absent
+            and not self._route_is_next_hop_to_auto_iface(route.to_dict())
         }
+        for ifname, routes in self.config_iface_routes.items():
```

Comment:
This for-loop seems to be focused on partial debugging (for desired routes that use a dynamic IP iface, also on the desired state).
I suggest this is not needed at the moment.
If it is needed, I suggest to group the absent routes and these next-hop-to-dynamic-iface into separate sets and use set operation and logging on them.

Reply:
Removed.